### ISSUE TITLE
Add ceiling fan support

### ISSUE THREAD

**Issue by sfussenegger** [frejos/openhab-wizlighting#3]
Wiz also supports (ceiling) fans with and without lights. Support should be added to control fan state, speed, direction as well as light.

Note: I'm a Java developer myself, so I'd be willing to put in some work here. I'm currently looking for ceiling fan option to integrate in my Openhab installation. I came across this [smart fan receiver from Faro Barcelona](https://faro.es/en/collection/smart-fan-receiver/) and was wondering how hard it would be to get this working (should I actually buy it). Currently I'm completely oblivious to Wiz though, so definitely would need some guidance.

**Comment by sfussenegger** [comment]
I've just came across this [discussion Reddit](https://www.reddit.com/r/wiz/comments/1bhhlj8/ceiling_fan/). Apparently fan support in Wiz itself is pretty new.

**Comment by frejos** [comment]
Hello,

I don't see any indication that Wiz has a ceiling fan module, nothing on the Wiz site about fan control modules. Also, the site you referenced does not indicate that the control modules offer a Wiz integration--I saw one comment by someone in the Reddit thread who stated that they did, but no collaboration of that statement that I've been able to find on the manufacture's site or data sheets for the modules.

All of the Wiz light products all pretty much work with the current binding configuration, along with the socket based products. These are detected as switches/dimmers as appropriate. If there were a ceiling fan module I would expect it to present itself as a dimmer and if so may work with the current binding. However, its also possible that a new module could use different commands or protocols than the existing Wiz components. The only way to tell for sure is to get one to try out.

Based off of what I could see in the data sheets (although they were not detailed), it doesn't look like these module are Wiz enabled modules. But if you get one and try it out and it turns out to be a Wiz based module, let me know.

**Comment by frejos** [comment]
There is however an option to add a Ceiling Fan in the Wiz app, which suggests that somewhere such a module exists.

**Comment by sfussenegger** [comment]
Thanks for the response and sorry for the late reply. I thought I'd just give it a go, order the fan and let you know as soon as I'm ready to play around with it. Just ordering that thing turned into a bit of an odyssey though. About 10 days ago I finally had one delivered but half-way through the installation I noticed that it actually wasn't the smart version. So that one was returned and now I'm back to waiting. I'm expecting the delivery next week but by then I'll be on vacation. As soon as I have it installed I'll let you know though.

**Comment by sfussenegger** [comment]
## So, finally ..

1.  I've got the fan installed and connected to Wiz
2. I've successfully [used `nc`](https://aleksandr.rogozin.us/blog/2021/8/13/hacking-philips-wiz-lights-via-command-line) to start and stop the fan
3. I've setup a dev environment with Eclipse, checked out the project and got it to build
4. I've installed the latest release of the addon which discovered (= Inbox) the Thing as a `wizDimmableBulb`

## nc commands

So at this point it might be good to get some assistance. Here are the `nc` commands I've used:

### get system config

```
echo '{"method":"getSystemConfig"}' | nc -u -w 1 192.168.0.167 38899 | jshon 
{
 "method": "getSystemConfig",
 "env": "pro",
 "result": {
  "mac": "d8a011c8xxxx",
  "homeId": 14692133,
  "roomId": 23105390,
  "rgn": "eu",
  "moduleName": "ESP03_FANDIMS_31",
  "fwVersion": "1.32.0",
  "groupId": 0,
  "ping": 0
 }
}
```

### get state

```
echo '{"method":"getPilot","params":{}}' | nc -u -w 3 192.168.0.167 38899 | jshon 
{
 "method": "getPilot",
 "env": "pro",
 "result": {
  "mac": "d8a011c8xxxx",
  "rssi": -76,
  "src": "",
  "state": true,
  "sceneId": 0,
  "dimming": 100,
  "fanState": 1,
  "fanSpeed": 6,
  "fanMode": 1,
  "fanRevrs": 0
 }
}
```

### turn off

```
echo '{"method":"setPilot","params":{"fanState":0}}' | nc -u -w 1 192.168.0.167 38899
{"method":"setPilot","env":"pro","result":{"success":true}}
```

### turn on

```
echo '{"method":"setPilot","params":{"fanState":1}}' | nc -u -w 1 192.168.0.167 38899
{"method":"setPilot","env":"pro","result":{"success":true}}
```

### set speed

```
echo '{"method":"setPilot","params":{"fanSpeed":6}}' | nc -u -w 1 192.168.0.167 38899
{"method":"setPilot","env":"pro","result":{"success":true}}
```

### reverse

```
echo '{"method":"setPilot","params":{"fanRevrs":1}}' | nc -u -w 1 192.168.0.167 38899
{"method":"setPilot","env":"pro","result":{"success":true}}
```

**Comment by frejos** [comment]
Awesome job figuring out the details to get this to work, at first glance it looks like you got all the key pieces. I will try to take a closer look at the pull request this weekend.

**Comment by frejos** [comment]
Does this module control a dimmable light as well as the fan? I'm trying to figure out why fanState is distinct from state and why it has fanSpeed and dimming? So state and dimming control the light fixture and fanState and fanSpeed control the fan? Otherwise if its just a fan control, It might make sense to push this into the existing channels.

**Comment by sfussenegger** [comment]
Ceiling fans come in variations with and without light. The one I have does not, so unfortunately I can't test it properly. It does however allow to set `state` and the Wiz app also discovers it as a fan with lighting, which suggests the controller does in fact support lighting (also explaining the unused connector cable coming it out of it). It seems to be dimmable too (as suggested by `dimming` property) but apparently not RGB. But that doesn't mean Wiz wouldn't support such models.

Also, I would have implemented the remaining channels fan speed, reverse and mode (though I don't really know [how that's different from reverse](https://wizconnected.helpshift.com/hc/en/7-wiz-v2/faq/776-fan-mode-introduction/?p=web)). Are you planning to do a release soon?

**Comment by sfussenegger** [comment]
Also the `Thread.sleep(..)` really isn't nice on the UI. I guess the proper solution would be to use `Registration` command on devices and handle `syncPilot` messages instead of immediately calling `getPilot` with the potential race condition. That's certainly some bigger change as it includes

1. setting up a UDP listening socket
2. managing registrations on devices (maybe just send it always before a command is sent? e.g. handling changed IP/mac address of openhab)
3. receiving, parsing and applying `syncPilot` messages from devices to the matching items

There'd be certainly some work to do but the big advantage would be that any external device interaction should be reflected in openHAB (e.g. my fan remote or using the Wiz mobile app). If you'd be open to this I'd add another ticket and might investigate whenever I find some time.

**Comment by frejos** [comment]
Yes, it looks like it does supports both a light and fan control. That was my next question, do you know the valid values for the fanSpeed and fanMode?

As for the `Thread.sleep(..)`, that's not a good solution. The binding already does the asynchronous listening for changes as you described. Periodically the binding requests the updated status from the device. The `getPilot()` function just sends the request to this device to update itself, the response is processed along with all the other update responses. Instead of doing a sleep, I changed this to schedule running the `getPilot()` function 1 second in the future.

I should be doing a release soon, but wanted to try to get these valid values in there if you know them.

**Comment by sfussenegger** [comment]
Ah, so you're doing the changes yourself. Fine with me. Scheduled update would have been my choice too. Just wasn't sure which way to go and went for the quick fix.

Just looked through the code again and found the `registration` stuff. I think I have to test it again with that in mind but it didn't seem to work for my fan. I'll have a look again and maybe come back with a bug report there.

As for the valid values: I don't know much about the `mode` values other than [the link I posted above](https://wizconnected.helpshift.com/hc/en/7-wiz-v2/faq/776-fan-mode-introduction/?p=web). I would have to do more testing, but apparently it's either `0` or `1` and it's basically just another way of setting `fanRevrs`. Not sure if other fans might have other modes or other behaviors with modes. In the docs of mine (if you even wanna call it that) I don't think it says something about modes other than reverse being "winter mode".

Regarding speed, mine is `1` to `6`. I would guess it's possible for other fans having more or less possible values though.
